### DISCUSSION TITLE
Having issue while creating multiple instances of a single provider used by a widget, and that widget is used multiple times on a single page.

### DISCUSSION THREAD

**Discussion by umersaddiq** [rrousselGit/riverpod#3536] [General]
Hi. I'm having a problem with using Riverpod

I have created reusable widget using generics, and to maintain their state, I have created a provider, as you can see in the code.

```
@Riverpod(keepAlive: false)
class RadioWidget<T> extends _$RadioWidget<T> {
  @override
  T? build() => null;

  void onChange(T? t) => state = t as T;
}

class AppRadioGroup<T> extends ConsumerStatefulWidget {
  const AppRadioGroup({super.key, required this.items});

  final List<T> items;

  @override
  ConsumerState<AppRadioGroup> createState() => _AppRadioGroupState();
}

class _AppRadioGroupState<T> extends ConsumerState<AppRadioGroup<T>> {
  @override
  Widget build(BuildContext context) {
    final value = ref.watch(radioWidgetProvider<T>());
    final notifier = ref.watch(radioWidgetProvider<T>().notifier);
    return Column(
      children: widget.items.map((e) {
        return RadioListTile<T>(
          value: e,
          groupValue: value,
          activeColor: Colors.red,
          onChanged: notifier.onChange,
          visualDensity: const VisualDensity(horizontal: -4, vertical: -4),
          title: Text(
            e.toString(),
            style: GoogleFonts.inter(color: Colors.white),
          ),
        );
      }).toList(),
    );
  }
}

enum Days { monday, tuesday, wednesday, thursday, friday, saturday, sunday }

enum Months { jan, feb, mar, apr, may, june, july, aug, sep, oct, nov, dec }

/// Usage

const AppRadioGroup<Days>(items: Days.values),
const AppRadioGroup<Months>(items: Months.values),

```

Now, the issue is that when I use this widget multiple times along with its provider, only a single instance of the provider gets created, which is then used by both widgets. Since I am assigning different Enums to T, why does only a single instance of the provider get created? How can I tackle this issue?

**Comment by gdurandrexel** [comment]
I suppose providers are not generic-aware.

Do you really need a provider to store the state of the widget? You could pass a [ValueNotifier](https://api.flutter.dev/flutter/foundation/ValueNotifier-class.html) as parameter to your widget, perhaps created with `useState` in the parent HookConsumerWidget.

Also the widget does not need to be stateful. A ConsumerWidget would suffice.

**Comment by umersaddiq** [comment]
@gdurandrexel riverpod version [3.0.0-dev.3](https://pub.dev/packages/riverpod/versions/3.0.0-dev.3) Supports generics. I want to do this using riverpod, and I don't want to use ValueNotifier or HookConsumerWidget.
